dhis-2/dhis-test-web-api/src/test/java/org/hisp/dhis/webapi/controller/DataElementControllerTest.java

```diff
@@ -133,4 +139,309 @@ void dataElementsExcludingDefaultCatComboTest() {
+  void prohibitValueTypeChangeWhenHasDataDeApiTest() {
+    // create metadata
+    POST(
+            "/metadata",
+            """
+            {
+                "dataElements": [
+                    {
+                        "id": "DeUid000001",
+                        "aggregationType": "DEFAULT",
+                        "domainType": "AGGREGATE",
+                        "name": "test de 1",
+                        "shortName": "test DE 1",
+                        "valueType": "NUMBER"
+                    }
+                ],
+                "organisationUnits": [
+                    {
+                        "id": "OrgUnitUid1",
+                        "name": "test org 1",
+                        "shortName": "test org 1",
+                        "openingDate": "2023-06-15"
+                    }
+                ],
+                "dataSets": [
+                    {
+                        "id": "DsUid000001",
+                        "name": "ds 1",
+                        "shortName": "ds 1",
+                        "periodType": "Monthly",
+                        "dataSetElements": [
+                            {
+                                "dataElement": {
+                                    "id": "DeUid000001"
+                                }
+                            }
+                        ]
+                    }
+                ]
+            }
+            """)
+        .content(OK);
+
+    // add org unit to user for data entry
+    PATCH(
+            "/users/" + ADMIN_USER_UID,
+            """
+            [
+                {
+                    "op": "add",
+                    "path": "/organisationUnits",
+                    "value": [
+                        {
+                            "id": "OrgUnitUid1"
+                        }
+                    ]
+                }
+            ]
+            """)
+        .content(OK);
+
+    // add data value for data element
+    POST(
+            "/dataValueSets",
+            """
+            {
+              "dataSet": "DsUid000001",
+              "period": "202311",
+              "orgUnit": "OrgUnitUid1",
+              "completedDate": "2023-11-05",
+              "dataValues": [
+                {
+                  "dataElement": "DeUid000001",
+                  "period": "202311",
+                  "orgUnit": "OrgUnitUid1",
+                  "value": "2000",
+                  "followup": false
+                }
+              ]
+            }""")
+        .content(OK);
+
+    // try update data element with new value type (TEXT)
+    JsonWebMessage validationErrorMsg =
+        assertWebMessage(
+            "Conflict",
+            409,
+            "ERROR",
+            "One or more errors occurred, please see full details in import report.",
+            PUT(
+                    "/dataElements/DeUid000001",
```

Comment:
added a test for both the `/dataElements` API and the `/metadata` API to cover different ways we allow updates and also in case the implementation of one diverges from the other.